dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/compilation/VisitorC11.java

```diff
@@ -160,41 +154,106 @@ public List<Event> visitAtomicXchg(AtomicXchg e) {
+    @Override
+    public List<Event> visitLlvmStore(LlvmStore e) {
+           return tagList(eventSequence(
+                newStore(e.getAddress(), e.getMemValue(), e.getMo())));
+    }
+
+    @Override
+    public List<Event> visitLlvmXchg(LlvmXchg e) {
+            Register resultRegister = e.getResultRegister();
+            ExprInterface value = e.getMemValue();
+            IExpr address = e.getAddress();
+            String mo = e.getMo();
+
+            Load load = newRMWLoadExclusive(resultRegister, address, mo);
+            Store store = newRMWStoreExclusive(address, value, mo, true);
+
+            return tagList(eventSequence(
+                            load,
+                            store));
+    }
+
+    @Override
+    public List<Event> visitLlvmRMW(LlvmRMW e) {
+            Register resultRegister = e.getResultRegister();
+            IOpBin op = e.getOp();
+            IExpr value = (IExpr) e.getMemValue();
+            IExpr address = e.getAddress();
+            String mo = e.getMo();
+
+            Register dummyReg = e.getThread().newRegister(resultRegister.getPrecision());
+            Local localOp = newLocal(dummyReg, new IExprBin(resultRegister, op, value));
+
+            Load load = newRMWLoadExclusive(resultRegister, address, mo);
+            Store store = newRMWStoreExclusive(address, dummyReg, mo, true);
+
+            return tagList(eventSequence(
+                            load,
+                            localOp,
+                            store));
+    }
+
+    @Override
+    public List<Event> visitLlvmCmpXchg(LlvmCmpXchg e) {
+            Register oldValueRegister = e.getStructRegister(0);
+            Register resultRegister = e.getStructRegister(1);
+
+            ExprInterface value = e.getMemValue();
+            IExpr address = e.getAddress();
+            String mo = e.getMo();
+            ExprInterface expectedValue = e.getExpectedValue();
+
+            Local casCmpResult = newLocal(resultRegister, new Atom(oldValueRegister, EQ, expectedValue));
+            Label casEnd = newLabel("CAS_end");
+            CondJump branchOnCasCmpResult = newJump(new Atom(resultRegister, NEQ, IValue.ONE), casEnd);
+
+            Load load = newRMWLoadExclusive(oldValueRegister, address, mo);
+            Store store = newRMWStoreExclusive(address, value, mo, true);
+
+            return tagList(eventSequence(
+                            // Indentation shows the branching structure
+                            load,
+                            casCmpResult,
+                            branchOnCasCmpResult,
+                                store,
+                            casEnd));
+    }
+
+    @Override
+    public List<Event> visitLlvmFence(LlvmFence e) {
+            return tagList(eventSequence(
+                            newFence(e.getMo())));
+    }
+
+    private List<Event> tagList(List<Event> in) {
+        return in.stream().map(e -> tagEvent(e)).collect(Collectors.toList());
+    }
```

Comment:
You shouldn't abuse `map` like this.
You can just do `in.forEach(this::tagEvent); return in;` and define `tagEvent` to not have a return value.